lib/sqlalchemy/dialects/mysql/expression.py

```diff
@@ -0,0 +1,148 @@
+from sqlalchemy.sql import elements
+from sqlalchemy.sql import operators
+
+
+def property_enables_flag(flag_name):
+    def wrapper(target):
+        @property
+        @wraps(target)
+        def inner(self):
+            new_flags = self.flags.copy()
+            new_flags[flag_name] = True
+
+            return match_(
+                self.clause,
+                against=self.against,
+                flags=new_flags,
+            )
+
+        return inner
+
+    return wrapper
+
+
+class match_(elements.ColumnElement):
+    """Produce a ``MATCH (X, Y) AGAINST ('TEXT')`` clause.
+
+    E.g.::
+
+        from sqlalchemy import desc
+        from sqlalchemy.mysql.dialects.mysql import match_
+
+        match_expr = match_(
+            users_table.c.firstname,
+            users_table.c.lastname,
+            against="John Connor",
+        )
+
+        stmt = select(users_table)\
+            .where(match_expr.in_boolean_mode)\
+            .order_by(desc(match_expr))
+
+    Would produce SQL resembling::
+
+        SELECT id, firstname, lastname
+        FROM user
+        WHERE MATCH(firstname, lastname) AGAINST (:param_1 IN BOOLEAN MODE)
+        ORDER BY MATCH(firstname, lastname) AGAINST (:param_2) DESC
+
+    The :func:`.match_` function is a standalone version of the
+    :meth:`_expression.ColumnElement.match_` method available on all
+    SQL expressions, as when :meth:`_expression.ColumnElement.match_` is
+    used, but allows to pass multiple columns
+
+    All positional arguments passed to :func:`.match_`, typically should be a
+     :class:`_expression.ColumnElement` instances
+
+    :param: against typically scalar expression to be coerced into a ``str``
+
+    :param: flags optional ``dict``. Use properties ``in_boolean_mode``,
+     ``in_natural_language_mode`` and ``with_query_expansion`` to control it::
+
+        match_expr = match_(
+            users_table.c.firstname,
+            users_table.c.lastname,
+            against="John Connor",
+        )
+
+        print(match_expr)
+
+        # MATCH(firstname, lastname) AGAINST (:param_1)
+
+        print(match_expr.in_boolean_mode)
+
+        # MATCH(firstname, lastname) AGAINST (:param_1 IN BOOLEAN MODE)
+
+        print(match_expr.in_natural_language_mode.with_query_expansion)
+
+        # MATCH(firstname, lastname) AGAINST
+        # (:param_1 IN NATURAL LANGUAGE MODE WITH QUERY EXPANSION)
+
+    :property: ``in_boolean_mode`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_boolean_mode`` flag
+
+    :property: ``in_natural_language_mode`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_natural_language`` flag
+
+    :property: ``with_query_expansion`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_query_expansion`` flag
+
+    .. versionadded:: 1.4.20
+
+    .. seealso::
+
+        :meth:`_expression.ColumnElement.match_`
+
+    """
+
+    default_flags = {
```

Comment:
not sure if maybe we could use immutable dict here

Reply:
@CaselIT good idea. I was on point python has no frozendict implementation. But [PEP416](https://www.python.org/dev/peps/pep-0416/) tells SQLAlchemy has own immutabledict implementation. In this case we need exactly immutability.

But in the same moment I have another idea. Is it better to make this class attribute private?
Or make it both?

Reply:
Yes sorry. Sqlalchemy has an immutabledict implementation that supports `union()` to create a new instance with the union of the options

I think it's ok to keep the parameter public, but it would be also fine if it's private (ie starts with underscore)

Regarding immutable container as you prefer. I don't think it's used for other similar constructs